render-store/service-worker.js

```diff
@@ -0,0 +1,57 @@
+/* global fetch */
+
+// Install the Service Worker ASAP.
+self.oninstall = function(event) {
+  event.waitUntil(self.skipWaiting());
+};
+
+self.onactivate = function(event) {
+  event.waitUntil(self.clients.claim());
+};
+
+// When fetching, distinguish on the method. This is naive but it suffices for
+// the example. For more sophisticated routing alternatives, use
+// [ServiceWorkerWare](https://github.com/gaia-components/serviceworkerware/)
+// or [sw-toolbox](https://github.com/GoogleChrome/sw-toolbox).
+self.onfetch = function(event) {
+  // For this example, `GET` implies looking for a cached copy...
+  if (event.request.method === 'GET') {
+    event.respondWith(getFromRenderStoreOrNetwork(event.request));
+  // While `PUT` means to cache contents...
+  } else {
+    event.respondWith(cacheInRenderStore(event.request).then(createdResponse));
+  }
+};
+
+// It tries to recover a cached copy for the document. If not found,
+// it respond from the network.
+function getFromRenderStoreOrNetwork(request) {
+  return self.caches.open('render-store').then(function(cache) {
+    return cache.match(request).then(function(match) {
+      return match || fetch(request);
+    });
+  });
+}
+
+// Obtains the interpolated HTML contents of a `PUT` request from the
+// `pokemon.js` client code and crafts an HTML response for the interpolated
+// result.
+function cacheInRenderStore(request) {
+  return request.text().then(function(contents) {
+    // Craft a `tesxt/html` response for the contents to be cached.
+    var headers = { 'Content-Type': 'text/html' };
+    var response = new Response(contents, { headers: headers });
+    return self.caches.open('render-store').then(function(cache) {
+      // Associate the crafted response with the
+      // [`referrer`](https://developer.mozilla.org/en-US/docs/Web/API/Request/referrer)
+      // property of the request which is the URL of the client page
+      // initiating the request.
+      return cache.put(request.referrer, response);
+    });
+  });
+}
+
+// Return a 202 response.
+function createdResponse() {
```

Comment:
Nit: since this function is used only once, you could avoid defining it here and just create an anonymous function where needed:
```
event.respondWith(cacheInRenderStore(event.request).then(function() {
  return new response
}));
```